WIP Split Crypto into standard Crypto and HeaderCrypto

yogeshsajanikar opened the pull request:
# Description

VRF and KES have new optimized implementations that need to be used for Conway era. This is the first step to use separate KES/VRF from Crypto such that it can be parametrized separately. Also since consensus is the primary consumer of the VRF/KES and ledger uses them for the bookkeeping, this PR moves HeaderCrypto into TPraos protocol.

This branch is based on a #lehins/no-vrf-usage-in-ledger (ported on the ledger release) to get conensus code compiled.

# Checklist

- [ ] Commit sequence broadly makes sense
- [ ] Commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-ledger/blob/master/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (which can be run with `scripts/fourmolize.sh`
- [x] Self-reviewed the diff

yogeshsajanikar committed:
WIP: Remove VRF references from PoolDistr

yogeshsajanikar committed:
WIP Split Crypto into Crypto and HeaderCrypto

This is still the work in progress and is rough at edges

- Split Crypto class into Crypto (DSIGN, ADDRHASH) and HeaderCrypto (KES, VRF)
- Move HeaderCrypto into Cardano.Protocol
- Use PoolStakeVRF and GenesisVRF as a proxy to VRF in ledger code
- Parametrize TPrao protcol code with Crypto and HeaderCrypto
  Many functions are parametrized by Era and HeaderCrypto as Era implicitly
  implies Crypto (Crypto Era).

TODO: Get examples and tests compiled

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Examples/Consensus.hs:
```diff
@@ -156,22 +163,22 @@ defaultShelleyLedgerExamples mkWitnesses mkAlonzoTx value txBody auxData transla
 exampleShelleyLedgerBlock ::
-  forall era.
-  (EraSegWits era, ShelleyBasedEra' era) =>
+  forall era hcrypto.
```
We've switched everywhere in the repo from `crypto` type parameter to `c`, at least that is the way it is now on master, so I think  it would be best to name it `hc`
```suggestion
  forall era hc.
```

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs:
```diff
@@ -431,15 +431,18 @@ validateGenesis
-              MaxKESEvolutionsUnsupported
-                sgMaxKESEvolutions
-                (totalPeriodsKES (Proxy @(KES (Crypto era))))
+      -- TODO: This is the only place where KES is required and that too for a
```
The whole `validateGenesis` function can be moved into `cardano-protocol-praos`, since it is not used anywhere in ledger.

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Examples/Consensus.hs:
```diff
@@ -567,5 +585,5 @@ examplePoolParams =
-  where
-    poolKeys = exampleKeys @c @'StakePool
+  -- where
+    -- poolKeys = exampleKeys @c @hc @'StakePool
```
```suggestion
```

lehins commented on eras/shelley/test-suite/src/Test/Cardano/Ledger/Shelley/Generator/Block.hs:
```diff
@@ -271,7 +284,7 @@ tickChainState
                 At (LastAppliedBlock {labHash}) -> hashHeaderToNonce labHash
             }
-        lv = either (error . show) id $ futureLedgerView testGlobals chainNes slotNo
+        -- lv = either (error . show) id $ futureLedgerView testGlobals chainNes slotNo
```
```suggestion
```

lehins commented on libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/API.hs:
```diff
@@ -245,7 +253,7 @@ instance CC.Crypto c => GetLedgerView (ConwayEra c) where
 
 -- | Data required by the Transitional Praos protocol from the Shelley ledger.
-data LedgerView crypto = LedgerView
+data LedgerView crypto hcrypto = LedgerView
```
Why do you need `hcrypto` in the `LedgerView`? It does not seem to being used.

yogeshsajanikar replied:
Not required. It is something that needs cleanup.

lehins reviewed (commented):
Yeah, this is looking good. I only had a couple of questions.

yogeshsajanikar committed:
WIP Separate KES VRF from Crypto

yogeshsajanikar committed:
WIP Separate header crypto

- ALl the ledger code is separated from header crypto (KES/VRF)
- Proxy VRFs are PoolStakeVRF and GenesisVRF

yogeshsajanikar committed:
Ormolise the souce

Format as per the style

yogeshsajanikar committed:
Fix ormolu errors

- Use Data.Proxy instead of a locally defined data type

yogeshsajanikar committed:
Update ruby version

ghost commented:
@lehins I am taking this over this PR from @yogeshsajanikar, as this is part of https://github.com/input-output-hk/ouroboros-network/issues/4151. Would you have some time to help me understand better what's the situation? I think I understand the overall goal of this work (eg. support evolving crypto primitives over eras) but it's not clear to me what's been done, what remains to be done, what's the deadline, whether there are intermediate steps we could do, etc.

lehins commented:
> It's not clear to me what's been done, what remains to be done, what's the deadline, whether there are intermediate steps we could do, etc.

@abailly-iohk I'll be happy to chat about it, however I am not too sure about the progress on this work or what the actual time line is. Considering that this PR is using 1.1 release branch as base this work is definitely outdated.

yogeshsajanikar commented:
The changes are done to the release branch as the main purpose was to

1 Split the crypto into two parts, header and body
2 Test these changes against consensus code.

Hence I changed release branch against which consensus code was compiled. Initially I targeted the ledger master branch, but the porting changes required in consensus were non-trivial.

The current status of the change is 

- the changes off the 1.1 release is compiled against consensus protocol library.

Having said that I agree that  work is needed to make it work against master. But most of the changes are mechanical.

ghost commented:
@yogeshsajanikar Thanks for your input, we have a chat today with @lehins to sort out this branch issue and move this PR forward.

ghost commented:
Had a shot at rebasing this issue on master and it's not trivial (for me), as discussed with @lehins I will wait for him to merge related changes to ledger's master codebase before trying again.

yogeshsajanikar commented:
Can I help? I have a some free time next week.

ghost commented:
@yogeshsajanikar Sure, would you be available on Monday morning CET for a call?

yogeshsajanikar commented:
Yes. Please feel free to schedule a meeting  @abailly-iohk

ghost commented:
Hope you're available, just setup a meeting today at 10:30 CET

ghost commented:
@lehins Do you have a branch we could attempt to rebase on?

lehins commented:
@abailly-iohk The one I have is quite unstable right this moment,  however I expect this to change either today or tomorrow. So, very soon I'll have a reviewable PR, which could be used as a base branch if need be.

ghost commented:
@lehins Any news on this front?

lehins commented:
@abailly-iohk Sorry, forgot to let you know. I did have a reviewable PR last week. As of today it has been merged, so https://github.com/input-output-hk/ouroboros-network/pull/4349 is fully compatible with ledger master. I'll be doing ledger releases to CHaPs today and merging the above PR

ghost commented:
No problem @lehins , so this means we can rebase this PR on master?

lehins commented:
> this means we can rebase this PR on master?

Correct

ghost commented:
@yogeshsajanikar Are you still available to do the rebase? Otherwise, I can take care of it, no worries.

yogeshsajanikar commented:
Yes, I can rebase it. Hope it's okay if I tackle it tomorrow? 

    On Wednesday, 22 February, 2023 at 08:46:36 pm IST, Arnaud Bailly ***@***.***> wrote:  
 
 


@yogeshsajanikar Are you still available to do the rebase? Otherwise, I can take care of it, no worries.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

ghost commented:
@yogeshsajanikar Tomorrow is totally fine, thanks for your help

yogeshsajanikar commented:
Started working on this.

yogeshsajanikar commented:
Continuing work today afternoon.

ghost commented:
@yogeshsajanikar I guess you did not have time to work on this? That's totally fine, I propose to take it over if you don't mind as I need it to implement the relevant stuff in consensus.

JaredCorduan commented:
This PR appears stale. I am going to close it for now.
